### ISSUE TITLE
Missing support for naming "XML array members"

### ISSUE THREAD

**Issue by jakubklimek** [opendata-mvcr/model-driven-data#126]
We are missing support for proper naming of "XML array members".
E.g. in Číselníky
![image](https://user-images.githubusercontent.com/718827/150762341-00489e3c-3364-4e1d-bf06-d2c68d314100.png)
We have "položky" as the technical label of the association (end).
However, that results in XML like this:
```xml
<číselník>
	<název>Můj číselník</název>
	<položky>
		<název>Položka 1</název>
	</položky>
	<položky>
		<název>Položka 2</název>
	</položky>
	<položky>
		<název>Položka 3</název>
	</položky>
</číselník>
```
whereas what we need is
```xml
<?xml version="1.0" encoding="UTF-8"?>
<číselník>
	<název>Můj číselník</název>
	<položky>
		<položka>
			<název>Položka 1</název>
		</položka>
		<položka>
			<název>Položka 2</název>
		</položka>
		<položka>
			<název>Položka 3</název>
		</položka>
	</položky>
</číselník>
```

We discussed that this could be solved by providing technical labels for both singular and plural, and then the XML generator would use those appropriately, while the JSON generator would ignore the singular.

However, we did not finish that discussion AFAIK.

**Comment by cermakmarek** [comment]
I think there is a conceptual difference here between a property with a list type and a property with unbounded cardinality. It seems to me that this distinction might also affect potential transformations from RDF:
```ttl
[
  a :číselník ;
  :položky (
    [ :název "Položka 1" ]
    [ :název "Položka 2" ]
    [ :název "Položka 3" ]
  )
] .
```
versus
```ttl
[
  a :číselník ;
  :položka [ :název "Položka 1" ] ;
  :položka [ :název "Položka 2" ] ;
  :položka [ :název "Položka 3" ] 
] .
```
Are both styles possible? Additionally, could the following situation happen as well?
```xml
<?xml version="1.0" encoding="UTF-8"?>
<číselník>
	<název>Můj číselník</název>
	<položky>
		<položka>
			<název>Položka 1</název>
		</položka>
		<položka>
			<název>Položka 2</název>
		</položka>
	</položky>
	<položky>
		<položka>
			<název>Jiná položka 1</název>
		</položka>
		<položka>
			<název>Jiná položka 2</název>
		</položka>
	</položky>
</číselník>
```
I.e. whether a property could both list-like and with higher cardinality.

**Comment by jakubklimek** [comment]
Yes, there is a difference. The first case represents an `rdf:List`, which is structurally a linked list implemented in RDF and is used to preserve ordering of items (a.k.a. [Ordered list](https://patterns.dataincubator.org/book/ordered-list.html) pattern). This would imply that the individual code list items have an ordering, which would otherwise be lost, when simply following the [Repeated property](https://patterns.dataincubator.org/book/repeated-property.html) pattern as in the second case. However, in this case, and IMHO even in general, we do not have a use case that would require the usage of `rdf:List` - we do not want to rely on it. One of the reasons is that it makes SPARQL querying harder, and it is usually simpler and more straightforward to use the [Ordering relation](https://patterns.dataincubator.org/book/ordering-relation.html).

In XML, the order of elements in element content is defined, validated in XML Schema and it is also defined in PSM, but here we are talking about the schema level, which is not present in RDF (unless talking about SHACL, etc., where ordering of the individual shapes is unimportant).

Even though on the instance level, the order of XML elements is also defined, and maybe someone would find it surprising to first see a codelist item "2" and then codelist item "1" in the XML file, in our use cases, it is unimportant at the moment.

The XML example you show should not be supported in XML -> RDF transformations.

**Comment by cermakmarek** [comment]
I see; yes, that makes sense then. I could make it so that if the maximum cardinality is higher than 1 and a singular name is specified, the element definition (with the singular property name and original cardinality) will be simply wrapped in another element with the plural name. Otherwise if cardinality is 0..1 or 1..1, it will be just a single element either with the singular name, if specified, or the original technical name. Is it correct?

Last question – should cardinality 0..* imply the optionality of `<položky>` itself? I.e. whether this is allowed:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<číselník>
	<název>Můj číselník</název>
</číselník>
```
instead of just an empty element:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<číselník>
	<název>Můj číselník</název>
	<položky/>
</číselník>
```

**Comment by jakubklimek** [comment]
I agree.

However, we should first agree on the singular/plural approach with @skodapetr and @martinnec - i.e. whether it is a specification-wide thing, or XML-specific one, etc.

I would expect the empty element to be there in this case, so that e.g. XPath queries explicitly return 0 elements instead of a missing element. However, it could get unnecessarily messy in sparse data with lots of `0..*` items.
@skodapetr how do you implement an optional JSON array? Do you have a preference (empty array vs. missing key and array)? 

As to the XML->RDF transformations, I think we should support both cases. It should not make a difference though - when I imagine an XSLT transformation - it will either not match the empty element at all, or it will, but will not generate anything.
